eth/stagedsync/stage_bodies.go

```diff
@@ -181,53 +226,109 @@ Loop:
 				cfg.bd.RequestSent(req, currentTime+uint64(timeout), peer)
 				d3 += time.Since(start)
 			}
+
+			loopCount++
+			if loopCount >= requestLoopCutOff {
+				break
+			}
 		}
-		start := time.Now()
-		headers, rawBodies, err := cfg.bd.GetDeliveries()
+
+		start = time.Now()
+		requestedLow, delivered, err := cfg.bd.GetDeliveries(innerTx)
 		if err != nil {
-			return err
+			return false, err
 		}
+		totalDelivered += delivered
 		d4 += time.Since(start)
 		start = time.Now()
-		cr := ChainReader{Cfg: cfg.chanConfig, Db: tx}
-		for i, header := range headers {
-			rawBody := rawBodies[i]
-			blockHeight := header.Number.Uint64()
-			// Txn & uncle roots are verified via bd.requestedMap
-			err := cfg.bd.Engine.VerifyUncles(cr, header, rawBody.Uncles)
-			if err != nil {
-				log.Error(fmt.Sprintf("[%s] Uncle verification failed", logPrefix), "number", blockHeight, "hash", header.Hash().String(), "err", err)
-				u.UnwindTo(blockHeight-1, header.Hash())
-				break Loop
+		cr := ChainReader{Cfg: cfg.chanConfig, Db: innerTx}
+
+		toProcess := cfg.bd.NextProcessingCount()
+
+		if toProcess > 0 {
+			var i uint64
+			for i = 0; i < toProcess; i++ {
+				nextBlock := requestedLow + i
+				key := dbutils.EncodeBlockNumber(nextBlock)
+				body, err := innerTx.GetOne("BodiesStage", key)
+				if err != nil {
+					return false, err
+				}
+				if body == nil {
+					return false, fmt.Errorf("[%s] Body was nil when reading from bucket, block: %v", logPrefix, nextBlock)
+				}
+
+				header, _, err := cfg.bd.GetHeader(nextBlock, cfg.blockReader, innerTx)
+				if err != nil {
+					return false, err
+				}
+
+				var rawBody types.RawBody
+				fromHex := common.CopyBytes(common.FromHex(string(body)))
+				bodyReader := bytes.NewReader(fromHex)
+				stream := rlp.NewStream(bodyReader, 0)
+				err = rawBody.DecodeRLP(stream)
+				if err != nil {
+					log.Error("Unexpected body from bucket", "err", err, "block", nextBlock)
+					return false, err
```

Comment:
If this info is useful - you can add it to error by: 
fmt.Errorf("%w, nextBlock=%d", err, nextBlock)